trafilatura/readability_lxml.py

```diff
@@ -440,3 +443,82 @@ def sanitize(self, node, candidates):
 
         self.doc = node
         return _tostring(self.doc)
+
+
+"""
+Port of isProbablyReaderable from mozilla/readability.js to Python.
+
+https://github.com/mozilla/readability
+
+License of forked code: Apache-2.0.
+"""
+
+REGEXPS = {
+    "unlikelyCandidates": re.compile(
+        r"-ad-|ai2html|banner|breadcrumbs|combx|comment|community|cover-wrap|disqus|extra|footer|gdpr|header|legends|menu|related|remark|replies|rss|shoutbox|sidebar|skyscraper|social|sponsor|supplemental|ad-break|agegate|pagination|pager|popup|yom-remote",
+        re.I,
+    ),
+    "okMaybeItsACandidate": re.compile(
+        r"and|article|body|column|content|main|shadow", re.I
+    ),
+}
+
+DISPLAY_NONE = re.compile(r"display:\s*none", re.I)
+
+
+def is_node_visible(node):
+    """
+    Checks if the node is visible by considering style, attributes, and class.
+    """
+    style = node.get("style")
+
+    if style and DISPLAY_NONE.search(style) is not None:
+        return False
+    if node.get("hidden") is not None:
+        return False
+    if node.get("aria-hidden") == "true" and "fallback-image" not in node.get(
+        "class", ""
+    ):
+        return False
+    return True
+
+
+def is_probably_readerable(html, options={}):
+    """
+    Decides whether or not the document is reader-able without parsing the whole thing.
+    """
+    doc = load_html(html)
+
+    min_content_length = options.get("min_content_length", 140)
+    min_score = options.get("min_score", 20)
+    visibility_checker = options.get("visibility_checker", is_node_visible)
+
+    nodes = doc.xpath(".//p | .//pre | .//article | .//div[.//br]")
```

Comment:
I'm not sure what `.//div[.//br]` does here, maybe you mean `.//div//br`?

Reply:
I tried to convert the CSS selector `"div > br"` to XPath, but it probably didn't work out. I think the correct XPath is `//div[br]`?

Reply:
Then I'd say it's `.//div//br` (br within div), the square brackets are for attributes if I'm not mistaken.

Reply:
I'm not really fluent with XPath, so I asked ChatGPT on the difference:

- `.//div[br]`: This XPath selects all `<div>` elements that contain at least one `<br>` element directly inside them. The `[br]` in the XPath is a condition that filters the `<div>` elements to only those that have a child `<br>`. It does not select the `<br>` elements themselves, but rather the `<div>` elements that meet this condition.

- `.//div//br`: This XPath selects all `<br>` elements that are descendants of a `<div>` element, regardless of how deep they are nested within the `<div>`. This could include `<br>` elements that are directly inside a `<div>`, as well as `<br>` elements that are nested further inside other child elements of the `<div>`. This XPath essentially traverses all levels of nested elements within each `<div>` to find `<br>` elements.

The original CSS selector was a [child selector](https://developer.mozilla.org/en-US/docs/Web/CSS/Child_combinator) with `div > br`:
> The child combinator matches only those elements matched by the second selector that are the direct children of elements matched by the first.

The square brackets are predicates that contain normal syntax. The attribute selector uses an `@` symbol like `.//div//[@attribute="value"]`. 

Here are some example s that also use this syntax:
> /PROJECT[STATUS]: Selects all projects that have a STATUS child element.
https://docs.oracle.com/javase%2Ftutorial%2F/jaxp/xslt/xpath.html

So I would say `.//div[br]` should be the right conversion from css `div > br`?

Reply:
For direct children I would use `//div/br`, maybe there are two different ways to write it.

I wouldn't trust GPT models on questions like this. Let's write a test for it if it's not already the case, that will settle the discussion.

Reply:
Here is result:
- `.//div[br]`: http://xpather.com/x39v3Sab
- `//div/br`: http://xpather.com/qgAWnSEC
- `.//div//br`: http://xpather.com/zL4D5mR5

So `//div/br` selects the `<br>` elements while `.//div[br]` selects the `<div>` elements. I also noticed my mistake when porting from JS to Python and why I had problems with the XPath selector in the first place:

https://github.com/mozilla/readability/blob/97db40ba035a2de5e42d1ac7437893cf0da31d76/Readability-readerable.js#L64-L71

This CSS selector selects the `<br>` but it adds the parent element (`<div`>) to the nodes set. I kept overlooking this.

Thanks for be persistent on that!